// bitbucketjs/lib/snippet.js
'use strict';

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _createDecoratedObject(descriptors) { var target = {}; for (var i = 0; i < descriptors.length; i++) { var descriptor = descriptors[i]; var decorators = descriptor.decorators; var key = descriptor.key; delete descriptor.key; delete descriptor.decorators; descriptor.enumerable = true; descriptor.configurable = true; if ('value' in descriptor || descriptor.initializer) descriptor.writable = true; if (decorators) { for (var f = 0; f < decorators.length; f++) { var decorator = decorators[f]; if (typeof decorator === 'function') { descriptor = decorator(target, key, descriptor) || descriptor; } else { throw new TypeError('The decorator for method ' + descriptor.key + ' is of the invalid type ' + typeof decorator); } } } if (descriptor.initializer) { descriptor.value = descriptor.initializer.call(target); } Object.defineProperty(target, key, descriptor); } return target; }

var _urlJoin = require('url-join');

var _urlJoin2 = _interopRequireDefault(_urlJoin);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _decorators = require('./decorators');

var log = (0, _debug2['default'])('bitbucketjs:snippet');
var path = '/2.0/snippets';

function formify(req) {
  var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

  req.type('multipart/form-data');

  Object.keys(options).forEach(function (k) {
    if (k !== 'owner' && k !== 'files') {
      req.field(k, options[k]);
    }
  });

  if (options.files) {
    options.files.forEach(function (f) {
      req.attach('file', f);
    });
  };
}

module.exports = function (opts) {
  var request = opts.request;

  function pathFor(ownername, snippetID) {
    return (0, _urlJoin2['default'])(opts.apiRoot, path, ownername, snippetID);
  }

  return _createDecoratedObject([{
    key: 'fetch',
    value: function fetch(ownername, snippetID) {
      return request.get(pathFor(ownername, snippetID));
    }
  }, {
    key: 'create',
    decorators: [(0, _decorators.authenticated)(request)],
    value: function create(options) {
      var req = request.post((0, _urlJoin2['default'])(opts.apiRoot, path, options.owner));

      formify(req, options);
      return req;
    }
  }, {
    key: 'update',
    decorators: [(0, _decorators.authenticated)(request)],
    value: function update(ownername, snippetID, options) {
      var req = request.put(pathFor(ownername, snippetID));

      formify(req, options);
      return req;
    }
  }, {
    key: 'delete',
    decorators: [(0, _decorators.authenticated)(request)],
    value: function _delete(ownername, snippetID) {
      return request.del(pathFor(ownername, snippetID));
    }
  }]);
};
